### ISSUE TITLE
`shared_codebook=True` not yielding correct results in eval mode

### ISSUE THREAD

**Issue by heffernankevin** [lucidrains/vector-quantize-pytorch#162]
Hello! 

Recently when I have been training with the argument: `shared_codebook=True`, I noticed that during validation the losses are extremely high. When I then evaluated the model I saw that in `train` mode my predictions are very good, but in `eval` mode I seem to be predicting noise. 

My previous setup was the exact same except for `shared_codebook=False`, so this argument is the only delta I can think of which is affecting the model somehow in eval mode. 

Any ideas?

**Comment by heffernankevin** [comment]
@lucidrains I think I've found what's going on here. Since each quantizer updates the same shared codebook, then perhaps this dependency during training is needed during inference. In other words, quantizer 1 expects the (shared) codebook to be updated based on quantizer 0. In this sense, the codebook needs to be adaptive at inference time and not fixed.

**Comment by lucidrains** [comment]
hey Kevin

was just about to ask you if you had EMA updates turned on

i think the right thing to do in a shared codebook setting is to delay the EMA updates until after the input has passed through all the quantizers

which paper are you looking at that prompts you to try shared codebook? it is not often used

**Comment by heffernankevin** [comment]
Thanks for the quick reply @lucidrains! What would the right setting be in this instance when initialising the RVQ? Would it be this?
```
ResidualVQ(
  dim=dim,
  codebook_dim=codebook_dim,
  num_quantizers=num_quantizers,
  codebook_size=num_codebooks,
  shared_codebook=True,
  ema_update=False,
)
```

Or would I need to manually set: `ema_update=False` for all the layers except for the last one? 

> which paper are you looking at that prompts you to try shared codebook? it is not often used

Just trying out the idea! :)

**Comment by heffernankevin** [comment]
> @heffernankevin yup, your first example should turn off the ema updates!

but then would I need to manually perform the ema updates?

**Comment by lucidrains** [comment]
@heffernankevin gave it a shot in 1.17.5, let me know if something doesn't look right

**Comment by heffernankevin** [comment]
> @heffernankevin gave it a shot in 1.17.5, let me know if something doesn't look right

Wow @lucidrains that was fast! I noticed in your update the codebook only gets updated once? `first(self.layers)._codebook.update_ema()`. In this sense, the residuals from the subsequent quantizers would not have any effect on the data? Would it make sense to update all codebook taking into account all the layers? (apologies if I'm missing something)

**Comment by lucidrains** [comment]
@heffernankevin the numerator (`embed_avg`) and denominator (`cluster_size`) is being updated live, just that the `embed` (actual codes used for quantization) are not updated

**Comment by lucidrains** [comment]
@heffernankevin still need to handle the expiration of the codes though, think can also get around that by accumulating all the inputs across layers and passing it in through some function on `VectorQuantizer`

**Comment by lucidrains** [comment]
@heffernankevin hey Kevin, just knocked out the expiration logic

want to give it another try? that training / eval discrepancy should be lessened

**Comment by heffernankevin** [comment]
> @heffernankevin hey Kevin, just knocked out the expiration logic
> 
> want to give it another try? that training / eval discrepancy should be lessened

trying it out now!

**Comment by lucidrains** [comment]
@heffernankevin what are you training on? feel free to reach out internally (i'm under Pytorch team until next June)

**Comment by lucidrains** [comment]
@heffernankevin issue resolved?
